### PR TITLE
feat: add systemd unit for scriptless

### PR DESCRIPTION
**Pull request by lilypan26**
**What type of PR is this?**

**What this PR does / why we need it**:
This PR adds a systemd unit to trigger bootstrapping for scriptless VHDs. The systemd unit waits for cloud-init completion (which ensure customdata files, in this case the boostrapping contract in json format, are written to disk) and then runs the go binary which takes the contract to generate cse_cmd to kickstart the bootstrapping process.

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:

```
none
```

### CODE REVIEW

**Comment by lilypan26** [comment]
Successful e2e run: https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=104976927&view=results

Corresponding build: https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=104963277&view=results

File: e2e/opts.go [hunk 1]
```diff
@@ -5,7 +5,8 @@ import (
 )
 
 type scenarioRunOpts struct {
-	clusterConfig *Cluster
-	scenario      *Scenario
-	nbc           *datamodel.NodeBootstrappingConfiguration
+	clusterConfig      *Cluster
+	scenario           *Scenario
+	nbc                *datamodel.NodeBootstrappingConfiguration
+	isSelfContainedVHD bool
```
Comment (cameronmeissner) [inline e2e/opts.go:11]: nit: `scriptless bool`

File: e2e/scenario_test.go [hunk 2]
```diff
@@ -636,6 +636,31 @@ func Test_ubuntu2204(t *testing.T) {
 	})
 }
 
+func Test_ubuntu2204Scriptless(t *testing.T) {
+	RunScenario(t, &Scenario{
+		Description: "Tests that a node using the Ubuntu 2204 VHD can be properly bootstrapped",
```
Comment (cameronmeissner) [inline e2e/scenario_test.go:641]: can we update the description to reference that it's using scriptless?

File: e2e/vmss.go [hunk 3]
```diff
@@ -232,6 +249,94 @@ func getVmssName(t *testing.T) string {
...
 
+func getBaseVMSSModelSelfContained(name, sshPublicKey string, opts *scenarioRunOpts) armcompute.VirtualMachineScaleSet {
+	nbc, err := json.Marshal(baseNodeBootstrappingContract(config.Config.Location, opts))
+	if err != nil {
+		return armcompute.VirtualMachineScaleSet{}
+	}
+
+	nbcEntry := `#cloud-config
+
+write_files:
```
Comment (cameronmeissner) [inline e2e/vmss.go:260]: can we move this to `templates.go`?

File: e2e/vmss.go [hunk 4]
```diff
@@ -232,6 +249,94 @@ func getVmssName(t *testing.T) string {
 	return name
 }
 
+func getBaseVMSSModelSelfContained(name, sshPublicKey string, opts *scenarioRunOpts) armcompute.VirtualMachineScaleSet {
+	nbc, err := json.Marshal(baseNodeBootstrappingContract(config.Config.Location, opts))
+	if err != nil {
+		return armcompute.VirtualMachineScaleSet{}
```
Comment (cameronmeissner) [inline e2e/vmss.go:255]: let's log the error at least

File: e2e/vmss.go [hunk 5]
```diff
@@ -232,6 +249,94 @@ func getVmssName(t *testing.T) string {
...
+		Properties: &armcompute.VirtualMachineScaleSetProperties{
+			Overprovision: to.Ptr(false),
+			UpgradePolicy: &armcompute.UpgradePolicy{
+				Mode: to.Ptr(armcompute.UpgradeModeManual),
+			},
+			VirtualMachineProfile: &armcompute.VirtualMachineScaleSetVMProfile{
+				OSProfile: &armcompute.VirtualMachineScaleSetOSProfile{
+					ComputerNamePrefix: to.Ptr(name),
+					AdminUsername:      to.Ptr("azureuser"),
+					CustomData:         &encodedCustomData,
```
Comment (cameronmeissner) [inline e2e/vmss.go:283]: so I guess only customdata?

File: nbcparser/main.go [hunk 6]
```diff
@@ -1,40 +1,36 @@
...
-// to build: go build main.go.
-// to run: ./main testdata/test_nbc.json.
 //
 //nolint:gosec // generated cse_cmd.sh file needs execute permissions for bootstrapping
+const (
+	CSE_CMD = "cse_cmd.sh"
+)
+
 func main() {
 	if len(os.Args) < parser.MinArgs {
```
Comment (cameronmeissner) [inline nbcparser/main.go:20]: nit(non-blocking): would prefer we use stdlib `flags` or `cobra` for flags rather than raw argv, can do that in another PR though

File: nbcparser/main.go [hunk 7]
```diff
@@ -1,40 +1,36 @@
...
-		log.Fatal(err)
-		return
-	}
-	if err = os.WriteFile("cse_cmd.sh", []byte(cseCmd), 0655); err != nil {
-		log.Fatal(err)
-		return
+		fmt.Println(err)
+		os.Exit(1)
 	}
+	fmt.Println(cseCmd)
```
Comment (cameronmeissner) [inline nbcparser/main.go:35]: do we still need this?

File: nbcparser/main.go [hunk 8]
```diff
@@ -1,40 +1,36 @@
 package main
 
 import (
-	"log"
```
Comment (cameronmeissner) [inline nbcparser/main.go:4]: did you want to actually switch over to `fmt`? personally I'd prefer sticking with `log`

File: nbcparser/pkg/parser/templates/cse_cmd.sh.gtpl [hunk 9]
```diff
@@ -148,4 +148,7 @@ IS_KATA="{{.GetIsKata}}"
 ARTIFACT_STREAMING_ENABLED="{{.GetEnableArtifactStreaming}}"
 SYSCTL_CONTENT="{{getSysctlContent .CustomLinuxOsConfig.GetSysctlConfig}}"
 PRIVATE_EGRESS_PROXY_ADDRESS=""
+BOOTSTRAP_PROFILE_CONTAINER_REGISTRY_SERVER=""
+ENABLE_IMDS_RESTRICTION="false"
```
Comment (cameronmeissner) [inline nbcparser/pkg/parser/templates/cse_cmd.sh.gtpl:152]: non-blocking: we should add some automation/PR-time checks to make sure this is in-sync with what we have in cse_cmd.sh

File: parts/linux/cloud-init/artifacts/nbcparser.sh [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+set -x
+# TODO: propagate exit status to bootstrapper
+if [ ! -f /opt/azure/containers/nbc.json ]; then 
+    exit 1
+fi
+/opt/azure/containers/nbcparser /opt/azure/containers/nbc.json > cse_cmd.sh
```
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/nbcparser.sh:7]: `/opt/azure/containers/nbcparser` will produce logs to stdout/err right? should we pipe that output to a dedicated log file? maybe something like `/var/log/azure/aks/bootstrap.log`

File: parts/linux/cloud-init/artifacts/nbcparser.target
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/nbcparser.target:1]: nit: `bootstrap.target`/`provision.target`?

Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/nbcparser.target:1]: or maybe even `cloud-init-complete.target` at that point to make things ultra clear

File: vhdbuilder/packer/pre-install-dependencies.sh [hunk 11]
```diff
@@ -80,6 +80,10 @@ capture_benchmark "enable_modified_log_rotate_service"
 systemctlEnableAndStart sync-container-logs.service || exit 1
 capture_benchmark "sync_container_logs"
 
+# enable and start nbcparser.service
+systemctl enable nbcparser.target
```
Comment (cameronmeissner) [inline vhdbuilder/packer/pre-install-dependencies.sh:84]: should we just be consistent and use `systemctlEnableAndStart` like we do for other units? or did we decide that we only wanted to enable here?

File: parts/linux/cloud-init/artifacts/nbcparser.sh [hunk 12]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+set -x
+# TODO: propagate exit status to bootstrapper
```
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/nbcparser.sh:3]: I think we will eventually have a boiler-plate CSE script which would run a script to poll the status of this unit and return an exit code based on the corresponding status - that'll be the only way we can get exit code information back to the caller I believe

File: e2e/vmss.go (hunk 5 above)
Comment (lilypan26) [reply e2e/vmss.go:275]: yep, will add the cse script to poll bootstrap status in separate pr

File: nbcparser/main.go (hunk 6 above)
Comment (lilypan26) [reply nbcparser/main.go]: modified to use `flag`

File: nbcparser/main.go (hunk 8 above)
Comment (lilypan26) [reply nbcparser/main.go:4]: switched back to log

File: nbcparser/main.go (hunk 7 above)
Comment (lilypan26) [reply nbcparser/main.go]: Yes, I was printing the generate cse_cmd to stdout and then piping it into a file to be executed in the systemd unit. However, I decided to just trigger cse_cmd from the go binary itself, refer to recent commits.

File: vhdbuilder/packer/pre-install-dependencies.sh (hunk 11 above)
Comment (lilypan26) [reply vhdbuilder/packer/pre-install-dependencies.sh]: we just want to enable, the new service will automatically be started once cloud-init is complete

File: vhdbuilder/packer/test/linux-vhd-content-test.sh [hunk 13]
```diff
@@ -960,14 +960,70 @@ testNBCParserBinary () {
...
+    return 1
+  fi
+  echo "$test: nbcparser.sh exists at $script_path"
+
+  # is-enabled returns:
+  # 'enabled' if the service is enabled.
+  # empty string if the service is not installed.
+  # 'not-found' if the unit files are not present. Encountered with Ubuntu 24.04
+  echo "$test: Checking that $service_name is enabled"
+  local is_enabled=
```
[the remark this answers is not in the document]
Comment (lilypan26) [reply vhdbuilder/packer/test/linux-vhd-content-test.sh]: hmmm doesn't like that: 
image

File: parts/linux/cloud-init/artifacts/nbcparser.sh (hunk 12 above)
Comment (lilypan26) [reply parts/linux/cloud-init/artifacts/nbcparser.sh]: Yep, will add this in the next PR

File: e2e/scenario_test.go [hunk 14]
```diff
@@ -649,6 +649,31 @@ func Test_ubuntu2204(t *testing.T) {
 	})
 }
 
+func Test_ubuntu2204Scriptless(t *testing.T) {
+	RunScenario(t, &Scenario{
+		Description: "Tests that a node using the Ubuntu 2204 Scriptless VHD can be properly bootstrapped",
+		Config: Config{
+			Cluster: ClusterKubenet,
+			VHD:     config.VHDUbuntu2204Gen2Containerd,
+			BootstrapConfigMutator: func(nbc *datamodel.NodeBootstrappingConfiguration) {
```
Comment (r2k1) [inline e2e/scenario_test.go:658]: Does it do anything? It looks like `*datamodel.NodeBootstrappingConfiguration` isn't used in your changes.

File: nbcparser/pkg/parser/templates/cse_cmd.sh.gtpl (hunk 9 above)
Comment (lilypan26) [reply nbcparser/pkg/parser/templates/cse_cmd.sh.gtpl:152]: Yes good point, will add this as a followup task

File: parts/linux/cloud-init/artifacts/nbcparser.sh (hunk 10 above)
Comment (lilypan26) [reply parts/linux/cloud-init/artifacts/nbcparser.sh]: added!

File: e2e/scenario_test.go (hunk 14 above)
Comment (lilypan26) [reply e2e/scenario_test.go:658]: I use the existing nbc to populate the new contract in `baseNodeBootstrappingContract` in `templates.go`. However,  I think you're right that these overrides aren't doing anything.

File: nbcparser/main.go [hunk 15]
```diff
@@ -1,40 +1,55 @@
...
+	test := flag.Bool("test", false, "test mode")
+	flag.Parse()
+	if *configFile == "" {
+		log.Fatal("bootstrap-config is a required argument")
 	}
 
-	// Read in the JSON file
-	inputJSON, err := os.ReadFile(os.Args[1])
+	cseCmd := parseConfig(*configFile)
+	err := bootstrap(cseCmd, *test)
```
Comment (Devinwong) [inline nbcparser/main.go:24]: where do we actually call the bootstrap() with test=true?

File: nbcparser/main.go [hunk 16]
```diff
@@ -1,40 +1,55 @@
...
 	if err != nil {
 		log.Fatal(err)
-		return
 	}
+	log.Println("Provisioning completed successfully")
```
Comment (Devinwong) [inline nbcparser/main.go:28]: I could be wrong. Is it more accurate to say Provisioning request initiated(received)? We don't know if it's successful or failed until the scripts really complete, right?

File: e2e/scenario_test.go [hunk 17]
```diff
@@ -649,6 +649,31 @@ func Test_ubuntu2204(t *testing.T) {
...
+				nbc.ContainerService.Properties.CertificateProfile.ClientPrivateKey = "client cert private key"
+				nbc.ContainerService.Properties.ServicePrincipalProfile.Secret = "SP secret"
+			},
+			LiveVMValidators: []*LiveVMValidator{
+				mobyComponentVersionValidator("containerd", getExpectedPackageVersions("containerd", "ubuntu", "r2204")[0], "apt"),
+				mobyComponentVersionValidator("runc", getExpectedPackageVersions("runc", "ubuntu", "r2204")[0], "apt"),
+			},
+		},
+		Tags: Tags{
+			Scriptless: true,
```
Comment (Devinwong) [inline e2e/scenario_test.go:672]: Does that mean by setting this tag `Scriptless`, client like aks-rp can determine sending scriptless contract or the old CSE scripts at runtime? Or this tag is only for e2e test?

File: e2e/opts.go [hunk 18]
```diff
@@ -8,4 +8,5 @@ type scenarioRunOpts struct {
 	clusterConfig *Cluster
 	scenario      *Scenario
 	nbc           *datamodel.NodeBootstrappingConfiguration
+	scriptless    bool
```
Comment (anujmaheshwari1) [inline e2e/opts.go:11]: super duper nit: this feels somewhat incomplete to me, `scriptlessVHD` maybe makes more sense to properly highlight what the scenario option is.

taking the cue from clusterConfig instead of just Config. feel free to ignore, just my opinion.

File: nbcparser/main.go (hunk 15 above)
Comment (lilypan26) [reply nbcparser/main.go:24]: in `linux-vhd-content-test.sh` we invoke the binary to test that it runs but don't want to actually run cse

File: parts/linux/cloud-init/artifacts/cse_start.sh [hunk 19]
```diff
@@ -1,5 +1,7 @@
 CSE_STARTTIME=$(date)
 CSE_STARTTIME_FORMATTED=$(date +"%F %T.%3N")
+EVENTS_LOGGING_DIR=/var/log/azure/Microsoft.Azure.Extensions.CustomScript/events/
```
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/artifacts/cse_start.sh:3]: did we ever determine from Chase if this was actually created by cse and if creating this manually had any implications?

File: parts/linux/cloud-init/artifacts/bootstrap.sh [hunk 20]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+set -x
```
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/artifacts/bootstrap.sh:2]: nit: set -euxo pipefail

File: parts/linux/cloud-init/artifacts/cse_start.sh (hunk 19 above)
Comment (lilypan26) [reply parts/linux/cloud-init/artifacts/cse_start.sh:3]: no chase said there should be nothing wrong with creating the directory manually, and I confirmed that I am still able to see the cse status in kusto with scriptless vhds.

File: parts/linux/cloud-init/artifacts/cloud-init-complete.target [hunk 21]
```diff
@@ -0,0 +1,5 @@
+[Unit]
+Description=Custom Target for Cloud Init Completion
```
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/artifacts/cloud-init-complete.target:2]: we dont really need this, you can do a `cloud-init status --wait` to ensure that cloud-init has finished before running anything else

File: e2e/scenario_test.go (hunk 17 above)
Comment (lilypan26) [reply e2e/scenario_test.go:672]: No, this is just for e2e testing

File: nbcparser/main.go (hunk 15 above)
Comment (Devinwong) [reply nbcparser/main.go:24]: Could you point me where it is? 
The reason I was asking this is to understand if we really need this variable for test purpose (in the meantime it will increase chances for the parser to run in test mode unintentionally, causing provisioning failure in production), or we can simply mock a test function in the test codes.
But if it's really necessary, we still can go with it. No preference now.

File: e2e/opts.go (hunk 18 above)
Comment (cameronmeissner) [reply e2e/opts.go:11]: lol `scriptless` was my original suggestion, I think `scriptlessVHD` is fine too

File: nbcparser/main.go (hunk 16 above)
Comment (lilypan26) [reply nbcparser/main.go]: Right, im just going to remove the print statement altogether since it doesn't tell us much. The cse script to poll the status of the systemd unit will provide information on whether bootstrapping completed successfully.

File: e2e/opts.go (hunk 18 above)
Comment (lilypan26) [reply e2e/opts.go:11]: i think i prefer simplicity of `scriptless` and we apparently we might rebrand again XD so just going to keep as is

File: parts/linux/cloud-init/artifacts/cloud-init-complete.target (hunk 21 above)
Comment (lilypan26) [reply parts/linux/cloud-init/artifacts/cloud-init-complete.target]: was able to get it working without this target by using systemd dependencies with cloud-init.target and cloud-final.service :)

File: nbcparser/main.go (hunk 15 above)
Comment (Devinwong) [reply nbcparser/main.go:24]: That makes sense. Thanks for clarifying.

**Review by Devinwong — APPROVED** [review]
LGTM. Thanks for the persistence for this feature. "No pressure" 😆

**Review by anujmaheshwari1 — APPROVED** [review]
